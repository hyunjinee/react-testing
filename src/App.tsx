import styled from 'styled-components';
import Button from 'components/common/Button';
import Input from 'components/common/Input';
import TodoItem from 'components/TodoItem';

function App() {
  return (
    <Container>
      <Contents>
        <TodoItem label="추가된 할 일" />
        <InputContainer>
          <Input placeholder="할 일을 입력해주세요." />
          <Button label="추가" />
        </InputContainer>
      </Contents>
    </Container>
  );
}

const Container = styled.div`
  min-height: 100vh;
  /* background: red; */

  display: flex;
  align-items: center;
  justify-content: center;
`;

const Contents = styled.div`
  display: flex;
  flex-direction: column;
  background-color: #ffffff;

  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;

const InputContainer = styled.div`
  display: flex;
`;

export default App;
