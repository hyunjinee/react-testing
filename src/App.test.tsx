import 'jest-styled-components';
import { fireEvent, render, screen } from '@testing-library/react';
import App from 'App';

describe('<App/>', () => {
  it('renders component correctly', () => {
    const { container } = render(<App />);

    const todoList = screen.getByTestId('todoList');
    expect(todoList).toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-node-access
    expect(todoList.firstChild).toBeNull();

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요.');
    expect(input).toBeInTheDocument();
    const label = screen.getByText('추가');
    expect(label).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('adds and deletes todo items', () => {
    render(<App />);

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요.');
    const button = screen.getByText('추가');
    fireEvent.change(input, { target: { value: '아이폰 사기' } });
    fireEvent.click(button);

    const todoItem = screen.getByText('아이폰 사기');
    expect(todoItem).toBeInTheDocument();
    const deleteButton = screen.getByText('삭제');
    expect(deleteButton).toBeInTheDocument();

    const todoList = screen.getByTestId('todoList');
    expect(todoList.childElementCount).toBe(1);

    fireEvent.change(input, { target: { value: '갤럭시 사기' } });
    fireEvent.click(button);

    const todoItem2 = screen.getByText('갤럭시 사기');
    expect(todoItem2).toBeInTheDocument();
    expect(todoList.childElementCount).toBe(2);
    const deleteButtons = screen.getAllByText('삭제');
    fireEvent.click(deleteButtons[0]);

    expect(todoItem).not.toBeInTheDocument();
    expect(todoList.childElementCount).toBe(1);
  });

  it('does not add empty Todo', () => {
    render(<App />);

    const todoList = screen.getByTestId('todoList');
    const length = todoList.childElementCount;

    const button = screen.getByText('추가');
    fireEvent.click(button);

    expect(todoList.childElementCount).toBe(length);
  });

  it('loads localStorage data', () => {
    localStorage.setItem(
      'todoList',
      JSON.stringify(['아이폰 사기', '갤럭시 사기']),
    );
    render(<App />);

    expect(screen.getByText('아이폰 사기')).toBeInTheDocument();
    expect(screen.getByText('갤럭시 사기')).toBeInTheDocument();
    expect(screen.getAllByText('삭제')).toHaveLength(2);
  });
});
