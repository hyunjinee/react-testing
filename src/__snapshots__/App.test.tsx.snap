// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders component correctly 1`] = `
.c4 {
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #ffffff;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  padding: 20px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  font-size: 20px;
  color: #ffffff;
  background-color: #304ffe;
  width: 60px;
  height: 60px;
  border-radius: 50%;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:hover {
  background-color: #1e40ff;
}

.c5:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #1e40ff;
}

.c2 {
  padding: 20px;
  color: #ffffff;
  font-weight: 600;
  font-size: 20px;
}

.c0 {
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        할 일 목록
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
        data-testid="todoList"
      />
      <a
        class="c5"
        href="/add"
      >
        +
      </a>
    </div>
  </div>
</div>
`;
